syntax = "proto3";

package envoy.api.v2.core;
option go_package = "core";

import "google/protobuf/struct.proto";
import "google/protobuf/wrappers.proto";

import "validate/validate.proto";
import "gogoproto/gogo.proto";

option (gogoproto.equal_all) = true;

// [#protodoc-title: Common types]

// Identifies location of where either Envoy runs or where upstream hosts run.
message Locality {
  // Region this :ref:`zone <envoy_api_field_core.Locality.zone>` belongs to.
  string region = 1;

  // Defines the local service zone where Envoy is running. Though optional, it
  // should be set if discovery service routing is used and the discovery
  // service exposes :ref:`zone data <config_cluster_manager_sds_api_host_az>`,
  // either in this message or via :option:`--service-zone`. The meaning of zone
  // is context dependent, e.g. `Availability Zone (AZ)
  // <https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html>`_
  // on AWS, `Zone <https://cloud.google.com/compute/docs/regions-zones/>`_ on
  // GCP, etc.
  string zone = 2;

  // When used for locality of upstream hosts, this field further splits zone
  // into smaller chunks of sub-zones so they can be load balanced
  // independently.
  string sub_zone = 3;
}

// Identifies a specific Envoy instance. The node identifier is presented to the
// management server, which may use this identifier to distinguish per Envoy
// configuration for serving.
message Node {
  // An opaque node identifier for the Envoy node. This also provides the local
  // service node name. It should be set if any of the following features are
  // used: :ref:`statsd <arch_overview_statistics>`, :ref:`CDS
  // <config_cluster_manager_cds>`, and :ref:`HTTP tracing
  // <arch_overview_tracing>`, either in this message or via
  // :option:`--service-node`.
  string id = 1;

  // Defines the local service cluster name where Envoy is running. Though
  // optional, it should be set if any of the following features are used:
  // :ref:`statsd <arch_overview_statistics>`, :ref:`health check cluster
  // verification <config_cluster_manager_cluster_hc_service_name>`,
  // :ref:`runtime override directory <config_runtime_override_subdirectory>`,
  // :ref:`user agent addition <config_http_conn_man_add_user_agent>`,
  // :ref:`HTTP global rate limiting <config_http_filters_rate_limit>`,
  // :ref:`CDS <config_cluster_manager_cds>`, and :ref:`HTTP tracing
  // <arch_overview_tracing>`, either in this message or via
  // :option:`--service-cluster`.
  string cluster = 2;

  // Opaque metadata extending the node identifier. Envoy will pass this
  // directly to the management server.
  google.protobuf.Struct metadata = 3;

  // Locality specifying where the Envoy instance is running.
  Locality locality = 4;

  // This is motivated by informing a management server during canary which
  // version of Envoy is being tested in a heterogeneous fleet. This will be set
  // by Envoy in management server RPCs.
  string build_version = 5;
}

// Metadata provides additional inputs to filters based on matched listeners,
// filter chains, routes and endpoints. It is structured as a map from filter
// name (in reverse DNS format) to metadata specific to the filter. Metadata
// key-values for a filter are merged as connection and request handling occurs,
// with later values for the same key overriding earlier values.
//
// An example use of metadata is providing additional values to
// http_connection_manager in the envoy.http_connection_manager.access_log
// namespace.
//
// For load balancing, Metadata provides a means to subset cluster endpoints.
// Endpoints have a Metadata object associated and routes contain a Metadata
// object to match against. There are some well defined metadata used today for
// this purpose:
//
// * ``{"envoy.lb": {"canary": <bool> }}`` This indicates the canary status of an
//   endpoint and is also used during header processing
//   (x-envoy-upstream-canary) and for stats purposes.
message Metadata {
  // Key is the reverse DNS filter name, e.g. com.acme.widget. The envoy.*
  // namespace is reserved for Envoy's built-in filters.
  map<string, google.protobuf.Struct> filter_metadata = 1;
}

// MetadataMatcher provides a general interface to check if a given value is matched in
// :ref:`Metadata <envoy_api_msg_core.Metadata>`. It uses `filter` and `path` to retrieve the value
// from the Metadata and then check if it's matched to one of the specified values.
//
// An example use of MetadataMatcher is specifying additional metadata in envoy.filters.http.rbac to
// enforce access control based on dynamic metadata in a request.
message MetadataMatcher {
  // Specifies the value to match. Only primitive value is supported.
  message Value {
    // Specifies how to match a value. Only have effect on primitive value.
    // For non-primitive value, it's always not matched.
    oneof match_specifier {
      option (validate.required) = true;

      // If specified, value match will be performed based on null value. It's matched if and only
      // if the target value is a NullValue and this field is set to true.
      bool null_match = 1;

      // If specified, value match will be performed based on double value. It's matched if and only
      // if the target value is a double value and is equal to this field.
      double number_match = 2;

      // If specified, value match will be performed based on string value. It's matched if and only
      // if the target value is a string value and is equal to this field.
      //
      // Examples:
      //
      // * *abc* will only match to string value *abc*.
      string exact_match = 3;

      // If specified, value match will be performed based on the prefix of the string value.
      // It's matched if and only if the target value is a string value and has the prefix specified
      // in this field.
      // Note: empty prefix is not allowed, please use present_match instead.
      //
      // Examples:
      //
      // * *abc* will match to string value *abc.xyz*
      string prefix_match = 4 [(validate.rules).string.min_bytes = 1];

      // If specified, value match will be performed based on the suffix of the string value.
      // It's matched if and only if the target value is a string value and has the suffix specified
      // in this field.
      // Note: empty suffix is not allowed, please use present_match instead.
      //
      // Examples:
      //
      // * *abc* will match to string value *xyz.abc*
      string suffix_match = 5 [(validate.rules).string.min_bytes = 1];

      // If specified, value match will be performed based on bool value. It's matched if and only
      // if the target value is a bool value and is equal to this field.
      bool bool_match = 6;

      // If specified, value match will be performed based on whether the path is referring to a
      // valid primitive value in the metadata.
      // Note: If the path is referring to a non-primitive value, it's always not matched.
      bool present_match = 7;
    }
  }

  // Required. The filter name to retrieve the Struct from the Metadata.
  string filter = 1 [(validate.rules).string.min_bytes = 1];

  // Required. The multi-key path to retrieve the Value from the Struct.
  repeated string path = 2 [(validate.rules).repeated .min_items = 1];

  // Required. A set of values to match. The MetadataMatcher is matched if at least one value is
  // matched, in other words, it's matched with OR semantics.
  repeated Value values = 3 [(validate.rules).repeated .min_items = 1];
}

// Runtime derived uint32 with a default when not specified.
message RuntimeUInt32 {
  // Default value if runtime value is not available.
  uint32 default_value = 2;

  // Runtime key to get value for comparison. This value is used if defined.
  string runtime_key = 3 [(validate.rules).string.min_bytes = 1];
}

// Envoy supports :ref:`upstream priority routing
// <arch_overview_http_routing_priority>` both at the route and the virtual
// cluster level. The current priority implementation uses different connection
// pool and circuit breaking settings for each priority level. This means that
// even for HTTP/2 requests, two physical connections will be used to an
// upstream host. In the future Envoy will likely support true HTTP/2 priority
// over a single upstream connection.
enum RoutingPriority {
  DEFAULT = 0;
  HIGH = 1;
}

// HTTP request method.
enum RequestMethod {
  option (gogoproto.goproto_enum_prefix) = false;
  METHOD_UNSPECIFIED = 0;
  GET = 1;
  HEAD = 2;
  POST = 3;
  PUT = 4;
  DELETE = 5;
  CONNECT = 6;
  OPTIONS = 7;
  TRACE = 8;
}

// Header name/value pair.
message HeaderValue {
  // Header name.
  string key = 1;

  // Header value.
  //
  // The same :ref:`format specifier <config_access_log_format>` as used for
  // :ref:`HTTP access logging <config_access_log>` applies here, however
  // unknown header values are replaced with the empty string instead of `-`.
  string value = 2;
}

// Header name/value pair plus option to control append behavior.
message HeaderValueOption {
  // Header name/value pair that this option applies to.
  HeaderValue header = 1;

  // Should the value be appended? If true (default), the value is appended to
  // existing values.
  google.protobuf.BoolValue append = 2;
}

// Data source consisting of either a file or an inline value.
message DataSource {
  oneof specifier {
    option (validate.required) = true;

    // Local filesystem data source.
    string filename = 1 [(validate.rules).string.min_bytes = 1];

    // Bytes inlined in the configuration.
    bytes inline_bytes = 2 [(validate.rules).bytes.min_len = 1];

    // String inlined in the configuration.
    string inline_string = 3 [(validate.rules).string.min_bytes = 1];
  }
}

// Configuration for transport socket in :ref:`listeners <config_listeners>` and
// :ref:`clusters <config_cluster_manager_cluster>`. If the configuration is
// empty, a default transport socket implementation and configuration will be
// chosen based on the platform and existence of tls_context.
message TransportSocket {
  // The name of the transport socket to instantiate. The name must match a supported transport
  // socket implementation.
  string name = 1 [(validate.rules).string.min_bytes = 1];

  // Implementation specific configuration which depends on the implementation being instantiated.
  // See the supported transport socket implementations for further documentation.
  google.protobuf.Struct config = 2;
}
